x/ibc/03-connection/keeper/keeper.go

```diff
@@ -0,0 +1,183 @@
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/ibc/03-connection/types"
+	ics23 "github.com/cosmos/cosmos-sdk/x/ibc/23-commitment"
+	ibctypes "github.com/cosmos/cosmos-sdk/x/ibc/types"
+)
+
+// Keeper defines the IBC connection keeper
+type Keeper struct {
+	storeKey  sdk.StoreKey
+	cdc       *codec.Codec
+	codespace sdk.CodespaceType
+	prefix    []byte // prefix bytes for accessing the store
+
+	clientKeeper types.ClientKeeper
+}
+
+// NewKeeper creates a new IBC connection Keeper instance
+func NewKeeper(cdc *codec.Codec, key sdk.StoreKey, codespace sdk.CodespaceType, ck types.ClientKeeper) Keeper {
+	return Keeper{
+		storeKey:     key,
+		cdc:          cdc,
+		codespace:    sdk.CodespaceType(fmt.Sprintf("%s/%s", codespace, types.DefaultCodespace)), // "ibc/connections",
+		prefix:       []byte(types.SubModuleName + "/"),                                          // "connections/"
+		clientKeeper: ck,
+	}
+}
+
+// Logger returns a module-specific logger.
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s/%s", ibctypes.ModuleName, types.SubModuleName))
+}
+
+// GetConnection returns a connection with a particular identifier
+func (k Keeper) GetConnection(ctx sdk.Context, connectionID string) (types.ConnectionEnd, bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), k.prefix)
+	bz := store.Get(types.KeyConnection(connectionID))
+	if bz == nil {
+		return types.ConnectionEnd{}, false
+	}
+
+	var connection types.ConnectionEnd
+	k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &connection)
+	return connection, true
+}
+
+// SetConnection sets a connection to the store
+func (k Keeper) SetConnection(ctx sdk.Context, connectionID string, connection types.ConnectionEnd) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), k.prefix)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(connection)
+	store.Set(types.KeyConnection(connectionID), bz)
+}
+
+// GetClientConnectionPaths returns all the connection paths stored under a
+// particular client
+func (k Keeper) GetClientConnectionPaths(ctx sdk.Context, clientID string) ([]string, bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), k.prefix)
+	bz := store.Get(types.KeyClientConnections(clientID))
+	if bz == nil {
+		return nil, false
+	}
+
+	var paths []string
+	k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &paths)
+	return paths, true
+}
+
+// SetClientConnectionPaths sets the connections paths for client
+func (k Keeper) SetClientConnectionPaths(ctx sdk.Context, clientID string, paths []string) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), k.prefix)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(paths)
+	store.Set(types.KeyClientConnections(clientID), bz)
+}
+
+// addConnectionToClient is used to add a connection identifier to the set of
+// connections associated with a client.
+//
+// CONTRACT: client must already exist
```

Comment:
Should this check for client existence and return an error if no?

Reply:
The spec ensures this contract, but it's up to the SDK team

Reply:
the client existence is checked on the caller function